core/src/main/java/org/infinispan/filter/CollectionKeyFilter.java

```diff
@@ -33,7 +34,11 @@ public CollectionKeyFilter(Collection<? extends K> keys, boolean accept) {
 
    @Override
    public boolean accept(K key) {
-      return accept ? keys.contains(key) : !keys.contains(key);
+      // This filter is provided as callback to cache stores, so it should be
+      // able to take normal keys and verify them against seen keys which
+      // might be MarshalledValues.
+      boolean contains = keys.stream().filter(k -> k.equals(key) || MarshalledValue.unwrap(k).equals(key)).count() > 0;
```

Comment:
I'm a bit worried about performance here, as state transfer uses the `CollectionKeyFilter` and the collection is the data container itself. 

It's weird that `AdvancedCacheLoader.process()` can skip the unmarshalling for values, but not for keys... I'm not sure we can avoid unnecessary marshalling/unmarshalling with the current API, but I think keeping `keys.contains(key)` is more important.